### ISSUE TITLE
`counsel-rg` misinterpret input starting with hyphen '-'

### ISSUE THREAD

**Issue by Alexander-Shukaev** [abo-abo/swiper#1833]
![bug](https://user-images.githubusercontent.com/911281/49332703-5262f180-f5b2-11e8-8c73-41b86fcbf86e.gif)

Clearly both searches produce nonsense results as there actually supposed to be zero matches for these inputs in the repository.  I suspect that the problem is with the implementation of argument forwarding to the called `rg` process.  That is, usually such utilities provide a way to say "please, treat the next input string as a non-option regardless of it starting with hyphen `-`), e.g. by supplying double hyphen `--` before it.  I didn't check whether `rg` supports exactly this method or offers something else but it's definitely worth looking into.  Most likely, the same bug would be found for `ag` and/or other search tools interfaces.

**Comment by basil-conto** [comment]
> I suspect that the problem is with the implementation of argument forwarding to the called `rg` process.

Most likely, yes - there is a lot of room for improvement in this department.

> I didn't check whether `rg` supports exactly this method or offers something else but it's definitely worth looking into.

It does, as do most command-line utilities. It also supports the explicit `-e PATTERN` switch, as per `grep(1)`, and as seen in the default `counsel-grep-base-command`.

The problem is that `counsel-rg-base-command` is modelled after `counsel-ag-base-command`, where the format spec `%s` is replaced not only with the search pattern, but potentially with further arguments.

I'm not familiar enough with this area to propose a fix, but in the meantime have you tried customising `counsel-rg-base-command` to add the switch separator `--` or an explicit `-e`?

**Comment by Alexander-Shukaev** [comment]
I can try it out, sure.

**Comment by Alexander-Shukaev** [comment]
ggreer/the_silver_searcher/pull/672

**Comment by basil-conto** [comment]
> https://github.com/ggreer/the_silver_searcher/pull/672

What are you trying to say? Just because `rg` understands the `-e` switch doesn't mean `ag` does, too. Every utility is different, and `ag` is buggier than most when it commes to command-line arguments. Still, `ag` supports the standard `--` separator.

**Comment by Alexander-Shukaev** [comment]
That's true.  The point now is that I wanted to use `--` as `counsel-rg` goes via `counsel-ag` anyways but then I discovered all that convoluted code for managing `extra-args` and parsing for occurrence of `--` which drove me mad because after reading that code I arrived to the conclusion that merely adding `--` before `%s` for both `counsel-ag-base-command` and `counsel-rg-base-command` will not work in all cases and in particular when `extra-args` are given.

**Comment by basil-conto** [comment]
Note that both `counsel-ag` and `counsel-rg` by extension support ad-hoc switches since #1408. What this means is that you can interactively add the `--` or `-e` switch to the search string in the minibuffer. Detecting the presence of `--` is currently a bit buggy, but otherwise the functionality is there.

**Comment by Alexander-Shukaev** [comment]
I know that it's supported and that's what I mentioned, the implementation is convoluted and wrong.  it needs to be refactored and `--` should always be there by default, for example, like I proposed above.  And if there is `--` in `extra-args` too, then it should be treated again like I proposed above, part before `--` goes to the first `%s` and part after to the second `%s`.
